File: python/best_stock_profit.py
def max_profit(prices):
    if len(prices) < 2:
        return 0
    
    min_price = prices[0]
    max_profit = float('-inf')
    
    for price in prices[1:]:
        max_profit = max(max_profit, price - min_price)
        min_price = min(min_price, price)
    
    return max_profit

File: ruby/best_stock_profit.rb
# https://www.codewars.com/kata/58f174ed7e9b1f32b40000ec/train/ruby

def max_profit(prices)
  return 0 if prices.size < 2
  
  min_price_so_far = prices[0]
  max_profit = -Float::INFINITY
  
  prices[1..-1].each do |price|
    max_profit = [max_profit, price - min_price_so_far].max
    min_price_so_far = [min_price_so_far, price].min
  end
  
  max_profit
end